### PR TITLE
Implementation of pmuludq (SSE2 instruction)

### PR DESCRIPTION
**Pull request by davidbullado**
- Implement pmuludq
- Nasm automated  tests : Add eight 128-bit SSE registers from xmm0 to xmm8

(My previous PR cannot be reopened, this is why I opened a new one. Sorry for the inconvenience)

### CODE REVIEW

**Comment by BenNottelling** [comment]
for people that don't have a local copy of closure compiler:
`wget https://www.bennottelling.com/closure-compiler.zip`
`unzip closure-compiler.zip`
either way, here's a zip with the pre-built js files
[v86_build_pmuludq.zip](https://github.com/copy/v86/files/2386768/v86_build_pmuludq.zip)
Cheers!

File: src/instructions.js [hunk 1]
```diff
@@ -4665,7 +4665,38 @@ t[0xF3] = cpu => {
...
+    cpu.read_modrm_byte();
+
+    if((cpu.prefixes & (PREFIX_MASK_REP | PREFIX_MASK_OPSIZE)) == PREFIX_66)
+    {
+        // pmuludq xmm1, xmm2/m128
+        let source = cpu.read_xmm_mem128s();
+        let destination = cpu.read_xmm128s();
+
+        let result_low = cpu.do_mul32(destination[0],source[0]);
+        let result_high = cpu.do_mul32(destination[2],source[2]);
```
Comment (copy) [inline src/instructions.js]: I don't think this works as the two `do_mul32`s alias their results (`do_mul32` doesn't do a copy). Could you add a regression test for this case?

File: src/instructions.js [hunk 2]
```diff
@@ -4665,7 +4665,38 @@ t[0xF3] = cpu => {
...
+        );
+    }
+    else
+    {
+        // pmuludq mm1, mm2/m64
+        dbg_assert((cpu.prefixes & (PREFIX_MASK_REP | PREFIX_MASK_OPSIZE)) == 0);
+        let source64s = cpu.read_mmx_mem64s();
+        let destination_low = cpu.reg_mmxs[2 * (cpu.modrm_byte >> 3 & 7)];
+
+        let result = cpu.do_mul32(destination_low,source64s[0])
```
Comment (copy) [inline src/instructions.js]: Minor: Put a space before `,` (also above)

**Comment by copy** [comment]
The Closure Compiler issue is being tracked here: https://github.com/google/closure-compiler/issues/3072. Since their website still links to it I'll assume it'll be back soon.

**Comment by BenNottelling** [comment]
I'm curios, but why did you implement pmuludq first? XD

**Comment by davidbullado** [comment]
> I'm curious , but why did you implement pmuludq first? XD
Yeah, I recognize that it was not a very wise choice xD. To be honest, it was like an easy entry point in the project for me, and I learn a lot since :) But now I will continue following a logical order, starting with sse1 intructions.

**Comment by copy** [comment]
@davidbullado Thanks!

For full disclosure: Right now I'm working on a closed-source fork of v86 written in Rust that already has full support for SSE 1 and 2. I hope to eventually merge it back into v86, but I don't think that will happen soon. I'll leave it up to you if you want to implement more SSE instructions in pure JavaScript, but I'll gladly review and merge them into the project.

**Comment by copy** [comment]
> That's good news ! Do you intend to target webassembly, by any chance ?

Yes, exactly.

> Thank you. I think this will keep myself busy the coming days (or weeks). I was also asking myself if you have a preference about the size of the PR ?

Anything between one and twenty instructions per PR works.

**Comment by davidbullado** [comment]
> How can I take the unimplemented SSE instruction (like `t[0x10] = cpu => { cpu.unimplemented_sse(); };`) and find whichever instruction it is? I've thought about trying to learn about everything and implement some, but I don't know how to. @davidbullado what sources did you use?

+ There is a good summary that I found [in an old PR](https://github.com/copy/v86/pull/162), (ty AmaanC): http://ref.x86asm.net/geek32.html
+ That reference help me also a lot: https://c9x.me/x86/
+ And there is also the source code of VirtualBox that can be helpful.
+ And google of course :)

**Comment by BenNottelling** [comment]
Awesome! Thank you so much :)
Now to see if I can find the motivation to learn everything about implementing it XD

**Comment by AmaanC** [comment]
> How can I take the unimplemented SSE instruction (like `t[0x10] = cpu => { cpu.unimplemented_sse(); };`) and find whichever instruction it is? I've thought about trying to learn about everything and implement some, but I don't know how to. @davidbullado what sources did you use?

An easy way to find which instructions are unimplemented SSE instructions in the wild (for eg. when running Windows XP on v86) is to run v86 in debug mode (i.e. `debug.html`) and to leave the developer tools open - the [`unimplemented_sse`](https://github.com/copy/v86/blob/aafaa7c878573c9a99dcca09467185845af4b12f/src/cpu.js#L3268-L3273) function calls [`dbg_assert(false)`, which will eventually call `debugger`](https://github.com/copy/v86/blob/aafaa7c878573c9a99dcca09467185845af4b12f/src/log.js#L105-L129), which will pause the code in your developer tools - in the call stack there, just click on the `t.(anonymous function).cpu` (or whatever it is in your tools) and follow the code:

I just added a dummy `unimplemented_sse` call to demonstrate.

Of course, you could also just update the debugging code to actually dump the opcode too, which might be a nice PR to have and a fun learning exercise :)
(Hint: Figure out how `instruction_pointer` changes as instructions call `unimplemented_sse` - does it vary? How does `read_imm8` work? Remember if you do add the logging, keep it to DEBUG mode only!)

I'm occasionally on the v86 Gitter too, if you have any questions I can help with (just ping me so I see it).

**Comment by davidbullado** [comment]
> Awesome! Thank you so much :)
> Now to see if I can find the motivation to learn everything about implementing it XD
> edit: @davidbullado do you have a gitter or discord or something in case I have any questions?
Yes I do, you can talk to me via the gitter of the project :)

**Pull request merged**